File: leetcode/medium/MinimumArrayEnd.py
class Solution:
    def minEnd(self, n: int, x: int) -> int:
        # The problem asks to find the n-th smallest number 'y' such that (x & y) == x.
        # This condition implies that 'y' must have all the bits set that are set in 'x'.
        # The bits that are not set in 'x' can be either 0 or 1 in 'y'.
        # We are essentially looking for the (n-1)-th combination of these "free" bits.

        # Initialize 'ans' with 'x'. This ensures all bits set in 'x' are also set in 'ans'.
        ans = x
        # Decrement 'n' by 1 because 'x' itself is the 1st number that satisfies the condition.
        n -= 1

        # Iterate through bit positions from 0 (least significant) upwards.
        # We will use the bits of 'n' to fill the "free" bit positions in 'ans'.
        bit_pos = 0
        while n > 0:
            # Check if the current bit position in 'x' is 0.
            # If it's 0, this bit position is "free" and can be set based on 'n'.
            if (x >> bit_pos) & 1 == 0:
                # If the least significant bit of 'n' is 1, set the current bit position in 'ans' to 1.
                if (n & 1) == 1:
                    ans |= (1 << bit_pos)
                # Move to the next bit of 'n'.
                n >>= 1
            # Move to the next bit position for 'x' and 'ans'.
            bit_pos += 1
            
        return ans

File: leetcode/medium/MinimumArrayEnd.java
/**
 * Solves the "Minimum Number of Operations" problem (often titled "Minimum End").
 * * Problem Statement:
 * Find the n-th smallest integer y >= x such that (y & x) == x.
 * * Logic:
 * The condition (y & x) == x means y must act as a superset of x.
 * 1. All bits set to 1 in x must remain 1 in y.
 * 2. The bits set to 0 in x are "free positions" or "holes".
 * 3. To find the n-th smallest number, we effectively map the binary 
 * representation of (n - 1) into these "free positions" in increasing order of significance.
 */
class Solution {

    /**
     * @param n The rank (1-based) of the number to find.
     * @param x The base number which acts as a bitmask.
     * @return The n-th smallest number y satisfying the condition.
     */
    public long minEnd(int n, int x) {
        // We need to inject (n-1) into the zero-bits of x.
        // We use a long to prevent overflow since the result can exceed Integer.MAX_VALUE.
        long result = x;
        long remaining = n - 1;
        long mask = 1L; // Starts at the 0-th bit (1)

        // Iterate as long as we have bits from (n-1) left to place
        while (remaining > 0) {
            // Check if the current bit position in x is a "free spot" (0)
            if ((x & mask) == 0) {
                // If the current bit of the remaining count is 1, set this bit in the result
                if ((remaining & 1) == 1) {
                    result |= mask;
                }
                // Move to the next bit of (n-1) essentially "consuming" it
                remaining >>= 1;
            }
            
            // Move the mask to check the next bit position
            mask <<= 1;
        }

        return result;
    }
}